### DISCUSSION TITLE
Implementations of explicit wait.

### DISCUSSION THREAD

**Discussion by DamianSzp** [eliasnogueira/selenium-java-lean-test-architecture#259] [General]
Hi @eliasnogueira
Could you share some examples of how the explicit wait can be implemented or some thoughts what is the best way to do it to keep code clean?

**Comment by eliasnogueira** [comment]
Hi @DamianSzp,
In this project, the [AbstractPageObject](https://github.com/eliasnogueira/selenium-java-lean-test-architecture/blob/main/src/main/java/com/eliasnogueira/page/AbstractPageObject.java) class manage it automatically thought the [AjaxElementLocatorFactory](https://www.selenium.dev/selenium/docs/api/java/org/openqa/selenium/support/pagefactory/AjaxElementLocator.html) class which will wait for a certain period to the element to appear.

All the page objects inherit the AbstractPageObject class, so all the elements in each object will have this verification, creating an "automatic way to wait for elements". This will make your code cleaner.

If you need to manage any additional explicitly wait, I would create its instance in the `AbstractPageObject`, using it inside the page objects to wait for what's needed in two different ways: inner waits or external waits.

Inner wait for me is the waiting usage inside a page object method.
The external one is a method that will wait for something.

Example in the `AbstractPageObject` class
```java
public class AbstractPageObject {

    protected WebDriverWait wait;
    
    protected AbstractPageObject() {
        initElements(new AjaxElementLocatorFactory(DriverManager.getDriver(), configuration().timeout()), this);
        wait = new WebDriverWait(DriverManager.getDriver(), Duration.ofSeconds(configuration().timeout()));
    }
}
```

Examples of waiting (fake, just as an example
```java
// code ignored

// inner
public void fillForm() {
    wait.until(ExpectedConditions.elementToBeSelected(By.id("id")));
    name.fill("something");
    // more code
}

// external
public void waitUntilMessageIsDisplayed() {
    wait.until(ExpectedConditions.visibilityOfElementLocated(By.id("message")));
}


```

**Reply by DamianSzp** [reply]
Thank you @eliasnogueira
